pkg/controller/pipelinerun/controller.go

```diff
@@ -0,0 +1,241 @@
+func NewController(
+	k kubernetes.Interface,
+	c clientset.Interface,
+	l *zap.SugaredLogger,
+) controller.Interface {
+
+	//TODO: Read timeout from a global config
+	pipelineFactory := informers.NewSharedInformerFactory(c, time.Second*30)
+	runsInformer := pipelineFactory.Pipeline().V1alpha1().PipelineRuns()
+	informer := pipelineFactory.Pipeline().V1alpha1().Pipelines()
+
+	// Enrich the logs with controller name
+	l = l.Named(constants.PipelineRunAgent).With(zap.String(logkey.ControllerType, constants.PipelineRunAgent))
+
+	controller := &Controller{
+		kubeClientset:   k,
+		clientset:       c,
+		runsLister:      runsListerFunc(runsInformer),
+		pipelinesLister: pipelinesListerFunc(informer),
+		synced:          runsInformer.Informer().HasSynced,
+		workqueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PipelineRuns"),
+		recorder:        getRecordFunc(k, l),
+		logger:          l,
+	}
+
+	l.Info("Setting up event handlers")
+	// Set up an event handler for when PipelineRun resource is added
+	runsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueuePipelineRun,
+	})
+	return controller
+}
+
+// Run will set up the event handlers for types we are interested in, as well
+// as syncing informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	// Start the informer factories to begin populating the informer caches
+	c.logger.Info("Starting PipelineRun controller")
+
+	// Wait for the caches to be synced before starting workers
+	c.logger.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.synced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	c.logger.Info("Starting workers")
+	// Launch two workers to process Build resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	c.logger.Info("Started workers")
+	<-stopCh
+	c.logger.Info("Shutting down workers")
+
+	return nil
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	if err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		key, ok := obj.(string)
```

Comment:
i think we could still use some comments here explaining these lines, even if they aren't as huge as [the original comments](https://github.com/knative/build/blob/master/pkg/controller/build/controller.go#L193)

Reply:
This no longer exists.